Sage/carbon
# fix(multiselect): ensure value prop is always used when controlled

When a controlled component has a custom onchange that sometimes does not update the value prop, the
component was not respecting this and still updating its internal value, essentially reverting the
component to being uncontrolled. This refactor, as well as fixing the bug, treats the controlled and
uncontrolled cases differently when the value changes so there should be less risk of similar bugs
in future.

fixes #5259

### Proposed behaviour



The pills shown in MultiSelect should always reflect the value prop when the component is controlled.

### Current behaviour

When an element is selected or deselected, the values shown as selected in the pills were always updating accordingly, even when the component has a custom onChange handler that leaves the value prop as it was.



### Checklist




#### QA


### Additional context



### Testing instructions



The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

https://codesandbox.io/s/multi-select-demo-forked-edzzss?file=/src/MultiSelectDemo.jsx

(note this is not identical to the CSB linked in the Github issue above. I have rewritten it so onChange checks the event.target.value.length rather than selectedPills length. If left as before then although the fix in this PR has the intended behaviour when selecting a third item, it made removal of items stop working, as selectedValue would never be updated at all once it reached a length of 2 or more. This small change makes the example work as intended after the fix in this PR, but it is still broken on the latest released Carbon.)
